File: algorithms/sorting/mergesort.rb
require 'test/unit'
extend Test::Unit::Assertions

def mergesort(list)
	if list.length < 2
		return list
	else
		mid = (list.length/2).to_i
		left = mergesort(list.slice(0, mid))
		right = mergesort(list.slice(mid, list.length))
		return _mergesort(left, right)
	end
end

def _mergesort(left, right)
	newarr = []
	i = j = 0
	while(newarr.length < left.length + right.length)
		if left[i] < right[j]
			newarr << left[i]
			i += 1
		else
			newarr << right[j]
			j += 1
		end
		if i == left.length
			right.slice(j, right.length).each {|x| newarr << x}
		elsif j == right.length
			left.slice(i, left.length).each {|x| newarr << x}
		end
	end
	return newarr
end

# Test cases
# Odd number of elements
assert(mergesort([7,6,5,4,3,2,1]) == [1,2,3,4,5,6,7], "Sorry, it did not pass through")
# Even number of elements
assert(mergesort([6,5,4,3,2,1]) == [1,2,3,4,5,6], "Sorry, it did not pass through")
# Repetition of elements
assert(mergesort([6,5,4,3,2,1,6]) == [1,2,3,4,5,6,6], "Sorry, it did not pass through")

File: algorithms/sorting/mergesort.py
# Top Down 
def mergesort(list):
	if len(list) > 1:
		mid = len(list) // 2
		left = mergesort(list[:mid])
		right = mergesort(list[mid:])
		return _mergesort(left, right)
	else:
		return list


# Helper merge function
def _mergesort(left, right):
	i, j = 0, 0
	newarr = []
	while(len(newarr) < len(left) + len(right)):
		if left[i] < right[j]:
			newarr.append(left[i])
			i += 1
		else:
			newarr.append(right[j])
			j += 1
		if i == len(left) or j == len(right):
			newarr.extend(left[i:] or right[j:])
	return newarr


# Test cases
# Odd number of elements
assert mergesort([7,6,5,4,3,2,1]) == [1,2,3,4,5,6,7]
# Even number of elements
assert mergesort([6,5,4,3,2,1]) == [1,2,3,4,5,6]
# Repetition of elements
assert mergesort([6,5,4,3,2,1,6]) == [1,2,3,4,5,6,6]

